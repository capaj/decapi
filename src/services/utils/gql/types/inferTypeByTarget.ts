import 'reflect-metadata'

import {
  GraphQLString,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLScalarType,
  getNullableType,
  GraphQLNonNull
} from 'graphql'
import { GraphQLDateTime } from 'graphql-scalars'
import { Constructor, reflect, ReflectedLiteralRef } from 'typescript-rtti'
import { Literal } from 'typescript-rtti/dist/common'

// tslint:disable-next-line: use-primitive-type
export type ParsableScalar = String | Number | Boolean | Date

export function isParsableScalar(input: any): input is ParsableScalar {
  return [String, Number, Boolean, Date].includes(input)
}

export function mapNativeTypeToGraphQL(input: any): GraphQLScalarType {
  switch (input) {
    case String:
      return GraphQLString
    case Number:
      return GraphQLFloat
    case Boolean:
      return GraphQLBoolean
    case Date:
      return GraphQLDateTime
    default:
      throw new Error(`Could not parse native type to graphql: ${input}`)
  }
}

function rttiLiteralToGql(rttLiteral: ReflectedLiteralRef<Literal>) {
  let literalType

  if (rttLiteral.isBooleanLiteral()) {
    literalType = GraphQLBoolean
  } else if (rttLiteral.isStringLiteral()) {
    literalType = GraphQLString
  } else if (rttLiteral.isNumberLiteral()) {
    literalType = GraphQLFloat
  }

  return literalType
}

export function inferTypeByTarget(target: Constructor<Function>, key?: string) {
  if (!key) {
    return Reflect.getMetadata('design:type', target)
  }
  const reflected = reflect(target)
  const property = reflected.getProperty(key)
  const method = reflected.getMethod(key)

  let rtti
  if (property) {
    rtti = property.type
  } else if (method) {
    rtti = method.returnType
  } else {
    throw new Error('Could not find property or method')
  }
  let inferred

  if (rtti.isClass()) {
    inferred = rtti.as('class').class
  } else if (rtti.isGeneric()) {
    // isPromise doesn't work properly here
    // promise
    if (rtti.as('generic').typeParameters[0].isUnion()) {
      const unionTypes = rtti.as('generic').typeParameters[0].as('union').types

      if (unionTypes.length === 2) {
        const withoutEmpties = unionTypes.filter(
          (x) => !x.isNull() || !x.isUndefined()
        )

        if (withoutEmpties.length === 2) {
          // TODO: handle union of two types
          throw new Error('Cannot infer the type')
        }
        if (withoutEmpties[0] instanceof ReflectedLiteralRef) {
          const rttLiteral = withoutEmpties[0].as('literal')
          const literalType = rttiLiteralToGql(rttLiteral, literalType)
          return getNullableType(literalType)
        }

        return getNullableType(
          mapNativeTypeToGraphQL(withoutEmpties[0].as('class').class)
        )
      } else {
        throw new Error('Unsupported union type') // decapi doesn't support unions with more than 2 types
      }
    } else {
      inferred = rtti.as('generic').typeParameters[0].as('class').class
      return new GraphQLNonNull(mapNativeTypeToGraphQL(inferred)) // TODO we would like to return nullable when we can detect that this type is implicit, depends on: https://github.com/rezonant/typescript-rtti/issues/16
    }
  } else if (rtti.isArray()) {
    const elementType = rtti.as('array').elementType
    if (elementType.isClass()) {
      inferred = elementType.as('class').class
    } else if (elementType.isUnion()) {
      const withoutEmpties = elementType
        .as('union')
        .types.filter((x) => !x.isNull() && !x.isUndefined())

      inferred = withoutEmpties[0].as('class').class
      return [getNullableType(mapNativeTypeToGraphQL(inferred))]
    }

    return [new GraphQLNonNull(mapNativeTypeToGraphQL(inferred))]
  }
  // console.log('~ inferred', inferred)

  if (isParsableScalar(inferred)) {
    return mapNativeTypeToGraphQL(inferred)
  }
  return inferred
}
