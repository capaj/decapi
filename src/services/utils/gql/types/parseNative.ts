import 'reflect-metadata'

import {
  GraphQLString,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLScalarType
} from 'graphql'
import { GraphQLDateTime } from 'graphql-scalars'

// tslint:disable-next-line: use-primitive-type
export type ParsableScalar = String | Number | Boolean | Date

export function isParsableScalar(input: any): input is ParsableScalar {
  return [String, Number, Boolean, Date].includes(input)
}

export function parseNativeTypeToGraphQL(input: any): GraphQLScalarType {
  switch (input) {
    case String:
      return GraphQLString
    case Number:
      return GraphQLFloat
    case Boolean:
      return GraphQLBoolean
    case Date:
      return GraphQLDateTime
  }
}

export function inferTypeByTarget(target: Function, key?: string) {
  if (!key) {
    return Reflect.getMetadata('design:type', target)
  }

  const returnType = Reflect.getMetadata('design:returntype', target, key)
  if (returnType) {
    return returnType
  }

  return Reflect.getMetadata('design:type', target, key)
}
