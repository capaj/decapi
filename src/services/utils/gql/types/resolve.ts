import { isType, GraphQLType, GraphQLList, GraphQLNonNull } from 'graphql'

import { parseNativeTypeToGraphQL, isParsableScalar } from './parseNative'
import {
  enumsRegistry,
  unionRegistry,
  inputObjectTypeRegistry,
  objectTypeRegistry,
  compileInputObjectType,
  compileObjectType
} from '../../../../index'

import { Thunk } from '../../../types'
import { interfaceTypeRegistry } from '../../../../domains/interfaceType/interfaceTypeRegistry'

export function resolveType(
  someType: any,
  allowThunk = true,
  isArgument?: boolean
): GraphQLType {
  if (isType(someType)) {
    return someType
  }

  if (isParsableScalar(someType)) {
    return parseNativeTypeToGraphQL(someType)
  }

  if (Array.isArray(someType)) {
    return resolveListType(someType, isArgument)
  }

  if (enumsRegistry.has(someType)) {
    return enumsRegistry.get(someType)
  }

  if (unionRegistry.has(someType)) {
    return unionRegistry.get(someType)()
  }

  if (interfaceTypeRegistry.has(someType)) {
    return interfaceTypeRegistry.get(someType)()
  }

  if (isArgument && inputObjectTypeRegistry.has(someType)) {
    return compileInputObjectType(someType)
  }

  if (objectTypeRegistry.has(someType)) {
    return compileObjectType(someType)
  }

  if (someType === Promise) {
    return
  }

  if (!allowThunk || typeof someType !== 'function') {
    return
  }
  let type
  try {
    type = someType()
  } catch (err) {
    if (err.message.match(/cannot be invoked without/)) {
      // a common error where you forget to decorate the class used as a decapi type
      throw new Error(
        `Class ${someType} needs to be decorated with @ObjectType in order for it to be usable as a type of a decapi field`
      )
    } else {
      throw err
    }
  }

  return resolveType(type, false)
}

function resolveListType(input: any[], isArgument: boolean): GraphQLType {
  if (input.length !== 1) {
    return
  }
  const [itemType] = input

  const resolvedItemType = resolveType(itemType, true, isArgument)

  if (!resolvedItemType) {
    return
  }
  return new GraphQLList(new GraphQLNonNull(resolvedItemType))
}

export function resolveTypesList(types: Thunk<any[]>): GraphQLType[] {
  if (Array.isArray(types)) {
    return types.map((type) => {
      return resolveType(type)
    })
  }
  return types().map((type) => {
    return resolveType(type)
  })
}
