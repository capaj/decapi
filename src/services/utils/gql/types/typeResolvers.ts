import { isType, GraphQLType, GraphQLList, GraphQLNonNull } from 'graphql'

import { parseNativeTypeToGraphQL, isParsableScalar } from './parseNative'
import {
  enumsRegistry,
  unionRegistry,
  inputObjectTypeRegistry,
  objectTypeRegistry,
  compileInputObjectType,
  compileObjectType
} from '../../../../index'

import { Thunk } from '../../../types'
import { interfaceTypeRegistry } from '../../../../domains/interfaceType/interfaceTypeRegistry'

/**
 * @param input is a type provided in the decorator config
 * @param allowThunk
 * @param isArgument
 */
export function resolveType(
  input: any,
  allowThunk = true,
  isArgument?: boolean
): GraphQLType {
  if (isType(input)) {
    return input
  }

  if (isParsableScalar(input)) {
    return parseNativeTypeToGraphQL(input)
  }

  if (Array.isArray(input)) {
    return resolveListType(input, isArgument)
  }

  if (enumsRegistry.has(input)) {
    return enumsRegistry.get(input)
  }

  if (unionRegistry.has(input)) {
    return unionRegistry.get(input)()
  }

  if (interfaceTypeRegistry.has(input)) {
    return interfaceTypeRegistry.get(input)()
  }

  if (isArgument && inputObjectTypeRegistry.has(input)) {
    return compileInputObjectType(input)
  }

  if (objectTypeRegistry.has(input)) {
    return compileObjectType(input)
  }

  if (input === Promise) {
    return
  }

  if (!allowThunk || typeof input !== 'function') {
    return
  }

  return resolveType(input(), false)
}

function resolveListType(input: any[], isArgument: boolean): GraphQLType {
  if (input.length !== 1) {
    return
  }
  const [itemType] = input

  const resolvedItemType = resolveType(itemType, true, isArgument)

  if (!resolvedItemType) {
    return
  }
  return new GraphQLList(new GraphQLNonNull(resolvedItemType))
}

export function resolveTypesList(types: Thunk<any[]>): GraphQLType[] {
  if (Array.isArray(types)) {
    return types.map((type) => {
      return resolveType(type)
    })
  }
  return types().map((type) => {
    return resolveType(type)
  })
}
