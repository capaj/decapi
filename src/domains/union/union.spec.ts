import { graphql, GraphQLUnionType, printSchema } from 'graphql'
import {
  ObjectType,
  Union,
  Field,
  compileObjectType,
  SchemaRoot,
  Query,
  compileSchema
} from '../..'

import 'jest'
import { resolveType } from '../../services/utils/gql/types/typeResolvers'

class ConstructorAssigner<T = any> {
  constructor(parameters: Partial<T>) {
    Object.assign(this, parameters)
  }
}

@ObjectType()
class Sub1 extends ConstructorAssigner {
  @Field()
  bar1: string | null
}

@ObjectType()
class Sub2 extends ConstructorAssigner {
  @Field()
  bar2: number | null
}

@Union({ types: [Sub1, Sub2] })
class UnionType {}

const customTypeResolver = jest.fn((type) => Sub1)

@Union({
  types: [Sub1, Sub2],
  resolveTypes: customTypeResolver
})
class CustomUnionType {}

@ObjectType()
class Foo {
  @Field({ type: UnionType })
  bar: Sub1 | Sub2

  @Field({ type: CustomUnionType })
  baz: Sub1 | Sub2
}

describe('Unions', () => {
  it('Registers returns proper enum type', () => {
    const { bar } = compileObjectType(Foo).getFields()
    expect(bar.type).toEqual(resolveType({ runtimeType: UnionType }))
    expect(bar.type).not.toEqual(UnionType)
  })

  it('Properly resolves type of union', () => {
    const { bar } = compileObjectType(Foo).getFields()

    const unionType = bar.type as GraphQLUnionType

    expect(
      unionType.resolveType &&
        // @ts-expect-error 3/21/2022
        unionType.resolveType(new Sub1(), null, null, null)?.toString()
    ).toBe(resolveType({ runtimeType: Sub1 }).toString())
    expect(
      unionType.resolveType &&
        // @ts-expect-error 3/21/2022
        unionType.resolveType(new Sub2(), null, null, null)?.toString()
    ).toBe(resolveType({ runtimeType: Sub2 }).toString())
  })

  it('Properly resolves with custom type resolver', () => {
    const { baz } = compileObjectType(Foo).getFields()

    const unionType = baz.type as GraphQLUnionType

    expect(
      unionType.resolveType &&
        // @ts-expect-error 3/21/2022
        unionType.resolveType(new Sub2(), null, null, null)
    ).toBe(resolveType({ runtimeType: Sub1 }))
    expect(customTypeResolver).toBeCalled()
  })

  it('should work in a schema', async () => {
    @SchemaRoot()
    class FooSchema {
      @Query({ type: [UnionType] })
      aUnion() {
        return [
          new Sub1({
            bar1: 'bar1'
          }),
          new Sub2({ bar2: 12 })
        ]
      }
    }
    const schema = compileSchema(FooSchema)

    expect(printSchema(schema)).toMatchInlineSnapshot(`
      "type Query {
        aUnion: [UnionType!]!
      }

      union UnionType = Sub1 | Sub2

      type Sub1 {
        bar1: String
      }

      type Sub2 {
        bar2: Float
      }"
    `)

    const result = await graphql({
      schema,
      source: `
        {
          aUnion {
            ... on Sub1 {
              bar1
            }
            ... on Sub2 {
              bar2
            }
          }
        }
      `
    })

    expect(result.errors).toBeUndefined()
    expect(result.data?.castedQuery).toMatchSnapshot()
  })
})
