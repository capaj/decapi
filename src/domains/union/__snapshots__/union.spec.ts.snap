// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unions should work in a schema 2`] = `undefined`;
